### ISSUE TITLE
[Bug] tab completion slow with slow imports

### ISSUE THREAD

**Issue by greaber** [facebookresearch/hydra#934]
# 🐛 Bug
## Description

Bash tab completion is slow for me (over two seconds even with a trivial config and a moderately complex config adds another 600ms).  It seems that the reason is that it is loading my whole script, and some imports are slow.

## Checklist
- [ x] I checked on the latest version of Hydra
- [ x] I created a minimal repro

## To reproduce

** Minimal Code/Config snippet to reproduce **

```
import hydra
import torch


@hydra.main(config_path="conf", config_name="config")
def main():
    pass

if __name__ == '__main__':
    main()
```

`import torch` here is just an example of a slow import.

## System information
- **Hydra Version** :  1.0.0rc2
- **Python version** : 3.7.6
- **Virtual environment type and version** : conda 4.8.2
- **Operating system** : Ubuntu 18.04

I'm running on a fast machine with ssd, no NFS or anything.

**Comment by omry** [comment]
Thanks for the report @greaber.
As you have identified this as caused by your imports, what you can do to improve things is to delay those imports.

### original:
main.py
```python
import hydra
import torch


@hydra.main(config_path="conf", config_name="config")
def main():
    pass

if __name__ == '__main__':
    main()
```

## modified:
main.py
```python
import hydra

@hydra.main(config_path="conf", config_name="config")
def main(cfg):
    import train
    train.train(cfg)

if __name__ == '__main__':
    main()
```

train.py
```python
import torch

def train(cfg):
  pass
```

**Comment by greaber** [comment]
I see.  I guess it could be pretty nice if the completion could use some kind of cache so it doesn't have to do this analysis on every completion, but I understand that would take a significant effort to implement, and it's not something I have time to do.

**Comment by omry** [comment]
Yup, running some kind of daemon would be perfect to cache the resulting config object, but as you pointed out this is a major effort and I am not sure it's worth it as this time.

Closing for now.

**Comment by omry** [comment]
I revisited this, and I think the original persistent daemon/cache idea is the right direction. Delaying expensive imports can help today, but it does not solve the general problem of repeating application startup and config composition on every completion request.

I added a [design direction for a persistent tab-completion accelerator](https://github.com/facebookresearch/hydra/blob/main/docs/design/tab_completion_service.md).

The proposed flow is:

```text
reploy install hydra-completion-accelerator
```

When tab completion is activated for a Hydra app, the app registers with the shared accelerator. Subsequent completion requests use it automatically. The accelerator can support multiple Hydra apps, keep their initialized composition environments available without running their task functions, and cache recent base compositions. It should also respond safely to source and config changes.

One-shot completion remains the fallback when the accelerator is unavailable. This is a direction for after Hydra 1.4, not an implemented feature yet.
